File: backend/app/services/pipeline/orchestrator.py
"""Pipeline orchestrator - chains stages together."""
from typing import List, Dict, Any
from app.queue.models import Job
from app.services.pipeline.base import PipelineStage, PipelineContext
from app.services.pipeline.stages.fetch_diff import FetchDiffStage
from app.services.pipeline.stages.analyze import AnalyzeStage
from app.services.pipeline.stages.aggregate import AggregateStage
from app.services.pipeline.stages.publish import PublishStage
from app.exceptions import PipelineError


class PipelineOrchestrator:
    """
    Orchestrates security analysis pipeline stages.

    Pipeline: Fetch Diff → Analyze → Aggregate → Publish
    """

    def __init__(self):
        """Initialize pipeline with default stages."""
        self.stages: List[PipelineStage] = [
            FetchDiffStage(),
            AnalyzeStage(),
            AggregateStage(),
            PublishStage(),
        ]

    def execute(self, job: Job) -> Dict[str, Any]:
        """
        Execute full pipeline for a job.

        Args:
            job: Job with PR details

        Returns:
            Dictionary with pipeline results

        Raises:
            PipelineError: If any stage fails
        """
        context = PipelineContext(job)

        # Execute each stage in sequence
        for stage in self.stages:
            try:
                context = stage.execute(context)
            except Exception as e:
                raise PipelineError(
                    f"Stage {stage.get_name()} failed: {str(e)}",
                    {"job_id": job.job_id, "stage": stage.get_name()}
                )

        return context.metadata

    def add_stage(self, stage: PipelineStage) -> None:
        """Add a custom stage to pipeline."""
        self.stages.append(stage)

    def remove_stage(self, stage_name: str) -> None:
        """Remove a stage by name."""
        self.stages = [s for s in self.stages if s.get_name() != stage_name]

File: backend/src/services/pipeline/orchestrator.ts
import { Job } from '../../queue/models.js';
import { PipelineStage, PipelineContext } from './base.js';
import { FetchDiffStage } from './stages/fetch-diff.js';
import { AnalyzeStage } from './stages/analyze.js';
import { AggregateStage } from './stages/aggregate.js';
import { PublishStage } from './stages/publish.js';
import { PipelineError } from '../../exceptions.js';

/**
 * Orchestrates security analysis pipeline stages.
 * Pipeline: Fetch Diff → Analyze → Aggregate → Publish
 */
export class PipelineOrchestrator {
  private stages: PipelineStage[];

  constructor() {
    this.stages = [
      new FetchDiffStage(),
      new AnalyzeStage(),
      new AggregateStage(),
      new PublishStage(),
    ];
  }

  /**
   * Execute full pipeline for a job.
   */
  async execute(job: Job): Promise<Record<string, unknown>> {
    let context = new PipelineContext(job);

    for (const stage of this.stages) {
      try {
        context = await stage.execute(context);
      } catch (e) {
        throw new PipelineError(
          `Stage ${stage.getName()} failed: ${(e as Error).message}`,
          { jobId: job.jobId, stage: stage.getName() }
        );
      }
    }

    return context.metadata;
  }

  /** Add a custom stage to pipeline. */
  addStage(stage: PipelineStage): void {
    this.stages.push(stage);
  }

  /** Remove a stage by name. */
  removeStage(stageName: string): void {
    this.stages = this.stages.filter(s => s.getName() !== stageName);
  }
}
